### PR TITLE
Added support for HTTP connection reuse (keep-alive) and SSL session …

### PR DESCRIPTION
**Pull request by jbrule**
Added support for HTTP connection reuse (keep-alive) and SSL session reuse. In API call heavy workloads this can drastically improve performance. Defaults to disabled

### CODE REVIEW

**Comment by jessp01** [comment]
Hello @jbrule ,

Thank you for your contribution (as well as for other recent ones).
It looks good but I do ask that since the only difference between TestMainPersistent.php and run-persistent.php and their original, non-persistent counterparts is the `$client->setPersistConnection(true)` call, you add an optional flag to the original test scripts instead.

Thanks again,

**Comment by jbrule** [comment]
Can do. That you for the feedback :)

**Comment by jbrule** [comment]
@jessp01 Refactored. I however am not able to successfully get the Php5Zend or Php53 tests to run. It fails at the mixingservice step in the multirequest (fails without my changes too). Could just be my testing environment. 

```
Php5Zend
==================
Running "php" run.php
Kaltura session (ks) was generated successfully: NDEwNDQ1N2JmN2
service url: [https://www.kaltura.com]
curl: https://www.kaltura.com/api_v3/service/system/action/ping
post: {"format":"2","ignoreNull":true,"clientTag":"php5:20-07-02","apiVersion":"16.4.0","ks":"NDEwNDQ1N2JmN2YwNDM2Y2FlMjU1Zjk2YjkzYmMyYzU5MzgxMjI3NTsyOzE4MTcwO3Rlc3RVc2VyOw==","kalsig":"cc7060f4271cb8c271c0898653264c76"}
persistent: false
server: [ny-front-api104], session: []
result (serialized): <?xml version="1.0" encoding="utf-8"?><xml><result>1</result><executionTime>0.01606297492981</executionTime></xml>
execution time for [https://www.kaltura.com/api_v3/service/system/action/ping]: [0.46711206436157]
Running testMultiRequest
service url: [https://www.kaltura.com]
curl: https://www.kaltura.com/api_v3/service/multirequest
post: {"format":"2","ignoreNull":true,"clientTag":"php5:20-07-02","apiVersion":"16.4.0","0":{"service":"mixing","action":"add","mixEntry":{"objectType":"KalturaMixEntry","editorType":"1","name":".Net Mix"},"ks":"NDEwNDQ1NkzYmMyYzdkYzg0ZTAxZnwyNTgxMjIxOzI1ODEyMjE7MTU5MzgxMjI3NTsyOzE4MTcwO3Rlc3RVc2VyOw=="},"1":{"service":"media","action":"upload","ks":"NDEwNDQ1FlMjU1Zjk2YjkzYmMyYzdkYzg0ZTAxZnwyNTgxMjIxOzI1ODEyMjE7MTU5MzgxMjI3NTsyOzE4MTcwO3Rlc3RVc2VyOw=="},"2":{"service":"media","action":"addFromUploadedFile","mediaEntry":{"objectType":"KalturaMediaEntry","mediaType":"1","name":"Media Entry For Mix"},"uploadTokenId":"{2:result}","ks":"NDEwNDQ1N2JmN2YwNDM2MyYzdkYzg0ZTAxZnwyNTgxMjIxOzI1ODEyMjE7MTU5MzgxMjI3NTsyOzE4MTcwO3Rlc3RVc2VyOw=="},"3":{"service":"mixing","action":"appendMediaEntry","mixEntryId":"{1:result:id}","mediaEntryId":"{3:result:id}","ks":"NDEwNDQ1N2JmN2YwNDM2Y2FlMjU1Zjk2YjkzYxZnwyNTgxMjIxOzI1ODEyMjE7MTU5MzgxMjI3NTsyOzE4MTcwO3Rlc3RVc2VyOw=="},"kalsig":"3ddb17a2c92131791f7b937f00037cee"}
persistent: false
server: [ny-front-api12], session: [319694276]
result (serialized): <?xml version="1.0" encoding="utf-8"?><xml><result><item><objectType>KalturaMixEntry</objectType><hasRealThumbnail>1</hasRealThumbnail><editorType>1</editorType><dataContent>&lt;xml&gt;
  &lt;MetaData&gt;
    &lt;Publish&gt;
      &lt;Application&gt;&apos;QuickEdit&apos;&lt;/Application&gt;
    &lt;/Publish&gt;
    &lt;SeqDuration&gt;0&lt;/SeqDuration&gt;
    &lt;ShowVersion&gt;-1&lt;/ShowVersion&gt;
  &lt;/MetaData&gt;
  &lt;VideoAssets/&gt;
  &lt;AudioAssets/&gt;
  &lt;VoiceAssets/&gt;
  &lt;Plugins/&gt;
&lt;/xml&gt;
</dataContent><plays>0</plays><views>0</views><duration>0</duration><msDuration>0</msDuration><id>1_l2xawrmj</id><name>.Net Mix</name><conversionProfileId>10772171</conversionProfileId><rootEntryId>1_e2vry57j</rootEntryId><operationAttributes></operationAttributes><entitledUsersEdit></entitledUsersEdit><entitledUsersPublish></entitledUsersPublish><entitledUsersView></entitledUsersView><capabilities></capabilities><displayInSearch>1</displayInSearch></item><item><error><objectType>KalturaAPIException</objectType><code>INTERNAL_SERVERL_ERROR</code><message>Internal server error occurred</message><args></args></error></item></result><executionTime>1.610787153244</executionTime></xml>
execution time for [https://www.kaltura.com/api_v3/service/multirequest]: [2.6274180412292]
testMultiRequest failed with error: Error occurred: Internal server error occurred
```

**Comment by jessp01** [comment]
Hello @jbrule ,

Indeed, these units fail here: https://github.com/kaltura/KalturaGeneratedAPIClientsPHP53/blob/master/tests/Test/Zend2ClientTester.php#L89 (same issue for the `zend` client). This has nothing to do with your contribution.
We'll fix that, run through a nightly cycle and merge following that.

Thanks,

**Comment by jessp01** [comment]
Can be merged as far as I'm concerned. @erankor, @MosheMaorKaltura - any inputs?

**Comment by jbrule** [comment]
Upon further review I think the configuration object may be a better place for the getter and setter. Like `getCurlTimeout`,`setCurlTimeout`. Maybe `getCurlReuse()` `setCurlReuse(true)`. Since it is curl specific. `closeCurlHandle` would still need to be on the Client object to allow closing the handle if it was desired. Thoughts?

**Comment by jessp01** [comment]
Sounds good to me, @jbrule. Go ahead and push the change and I'll review.

**Comment by jbrule** [comment]
@jessp01 Configuration of setting has been moved into configuration object. I did add getter/setters to the configuration object across all SDKs to make consistent but it looks like the practice varys.

**Comment by jessp01** [comment]
Hi @jbrule,

Your changes look good. However:
- I don't think we need the `curlHandle: ` log message (I can understand how it was useful to you while doing the initial testing)
- The `persistent:` log message should be a bit clearer, IMHO, something along the lines of "Reusing existing cURL handle"
- Please document this option in the README.txt per client

Thanks,

**Comment by jbrule** [comment]
@jessp01 

Logging labeling improvement made. I'm seeking guidance on how the README.txt should be formatted to document this. Currently the README for the affected clients are only 2 lines.

**Comment by jessp01** [comment]
Hi @jbrule,

Indeed, the current README is quite bare and there's sadly no consistency between READMEs for the various clients.
I propose we use https://github.com/kaltura/clients-generator/blob/Propus-16.7.0/sources/node2/README.md as a baseline.
We can rename the existing README.txt to benefit from the markdown formatting.

File: sources/zend/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+## Kaltura PHP API Client Library for Zend framework.
+Compatible with Kaltura server version @VERSION@ and above.
+
+## Install
+TODO
+
+## Getting Started
+
+### Enumerating Content Library
```
Comment (jessp01) [inline sources/zend/README.md:9]: Let's reference https://github.com/kaltura/KalturaGeneratedAPIClientsZF/blob/master/tests/run.php, please.

**Comment by jessp01** [comment]
Hi @jbrule,

Thanks again for your contributions:) You did a fine job! Please see my latest comments.

As for the existing README.md, it was committed directly to the https://github.com/kaltura/KalturaGeneratedAPIClientsPHP repo. Since your version is far more comprehensive, I'm happy for it to be overridden.

**Comment by jessp01** [comment]
All looks good to me @jbrule.
@MosheMaorKaltura, @yossipapi - are we good to merge?

**Comment by jbrule** [comment]
@jessp01 @MosheMaorKaltura @yossipapi  Looks like this has lost momentum, probably due to attention to AWS migration, which I understand. I have an app which makes a lot of API calls and using this pretty much cut page generation time in half, so you make also see benefits.

**Comment by jbrule** [comment]
FYI. An interesting benchmark I thought you'd like to see. I have an script that finds 657 existing categories and media using three API calls for each (not using multirequest) for 1971 calls total. I ran them a couple times to balance out any differences due to mysql query cache or other caches.

I am using **https** serviceUrl.

Run with default (curlReuse not enabled)
`php process.php  16.50s user 1.96s system 3% cpu 7:56.28 total`

With curlReuse enabled
`php process.php  0.56s user 0.23s system 1% cpu 1:11.89 total`

**Comment by jbrule** [comment]
Ping

**Pull request closed without merging**